subprojects/base-services/src/main/java/org/gradle/internal/work/DefaultWorkerLeaseService.java

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.work;
+
+import com.google.common.collect.Lists;
+import org.gradle.api.Action;
+import org.gradle.api.Describable;
+import org.gradle.api.Transformer;
+import org.gradle.api.specs.Spec;
+import org.gradle.internal.resources.AbstractResourceLockRegistry;
+import org.gradle.internal.resources.AbstractTrackedResourceLock;
+import org.gradle.internal.resources.DefaultResourceLockCoordinationService;
+import org.gradle.internal.resources.ExclusiveAccessResourceLock;
+import org.gradle.internal.resources.ResourceLock;
+import org.gradle.internal.resources.ResourceLockCoordinationService;
+import org.gradle.internal.resources.ResourceLockState;
+import org.gradle.util.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.gradle.internal.resources.DefaultResourceLockCoordinationService.*;
+import static org.gradle.internal.resources.ResourceLockState.Disposition.*;
+
+public class DefaultWorkerLeaseService implements WorkerLeaseService {
```

Comment:
I think `WorkerLeaseService` is too specific for the now more generic nature of this interface. Maybe just `LeaseService`?